### PR TITLE
Bug 2105918, Bug 2105303: specify namespace and index entry along with chart url to get the chart details

### PR DESCRIPTION
**Pull request by debsmita1**
**Bugzilla:**
https://bugzilla.redhat.com/show_bug.cgi?id=2105303
https://bugzilla.redhat.com/show_bug.cgi?id=2105918

**Solution description:**
- passing namespace and chart index entry wherever chart URL is passed to get chart details, install chart or upgrade chart version

**Test coverage:**
added new tests

### CODE REVIEW

File: frontend/packages/helm-plugin/src/catalog/components/HelmReadmeLoader.tsx [hunk 1]
```diff
@@ -19,7 +25,9 @@ const HelmReadmeLoader: React.FC<HelmReadmeLoaderProps> = ({ chartURL }) => {
       let chartData;
 
       try {
-        chartData = await coFetchJSON(`/api/helm/chart?url=${chartURL}`);
+        chartData = await coFetchJSON(
+          `/api/helm/chart?url=${chartURL}&namespace=${namespace}&indexEntry=${chartIndexEntry}`,
+        );
```
Comment (jerolimov) [inline frontend/packages/helm-plugin/src/catalog/components/HelmReadmeLoader.tsx:30]: We need to escape at least `chartURL` here, right?

Namespace names are "safe" for any injection because of their limited character subset, but we might escape all parameters.

File: frontend/packages/helm-plugin/src/components/forms/install-upgrade/HelmChartVersionDropdown.tsx [hunk 2]
```diff
@@ -141,8 +150,9 @@ const HelmChartVersionDropdown: React.FunctionComponent<HelmChartVersionDropdown
 
     setFieldValue('chartVersion', value);
     setFieldValue('chartURL', chartURL);
-
-    coFetchJSON(`/api/helm/chart?url=${chartURL}`)
+    coFetchJSON(
+      `/api/helm/chart?url=${chartURL}&namespace=${namespace}&indexEntry=${chartIndexEntry}`,
+    )
```
Comment (jerolimov) [inline frontend/packages/helm-plugin/src/components/forms/install-upgrade/HelmChartVersionDropdown.tsx:155]: Same here, we need to escape at least `chartURL`. I would prefer if we escape all parameter values.

File: frontend/packages/helm-plugin/src/catalog/utils/catalog-utils.tsx [hunk 3]
```diff
@@ -43,7 +43,7 @@ export const normalizeHelmCharts = (
         const imgUrl = chart.icon || getImageForIconClass('icon-helm');
         const chartURL = chart.urls[0];
         const encodedChartURL = encodeURIComponent(chartURL);
-        const href = `/catalog/helm-install?chartName=${name}&chartRepoName=${chartRepoName}&chartURL=${encodedChartURL}&preselected-ns=${activeNamespace}`;
+        const href = `/catalog/helm-install?chartName=${name}&chartRepoName=${chartRepoName}&chartURL=${encodedChartURL}&preselected-ns=${activeNamespace}&indexEntry=${key}`;
```
Comment (jerolimov) [inline frontend/packages/helm-plugin/src/catalog/utils/catalog-utils.tsx:46]: Interessting that we encode chartURL here, but not in our API calls. And here we should also escape name and chartRepoName, right? Or are both kubernetes resources names? It might the safest way to encode them anyway.

File: frontend/packages/helm-plugin/src/utils/helm-utils.ts [hunk 4]
```diff
@@ -206,7 +222,7 @@ export const getHelmActionConfig = (
             'helm-plugin~The Helm Chart can be installed by manually entering YAML or JSON definitions.',
           ),
         },
-        helmReleaseApi: `/api/helm/chart?url=${chartURL}`,
+        helmReleaseApi: `/api/helm/chart?url=${chartURL}&namespace=${namespace}&indexEntry=${chartIndexEntry}`,
```
Comment (jerolimov) [inline frontend/packages/helm-plugin/src/utils/helm-utils.ts:225]: Here we should also encode the chartURL.

**Review by jerolimov — CHANGES_REQUESTED** [review]
The chartURL is encoded in `packages/helm-plugin/src/catalog/utils/catalog-utils.tsx` and IMHO we should do this on all API calls and links. And its btw something we should also do for all parameters from an "unknown" source. I would accept that we don't do it for k8s resource names because of their limited charset.

Let me know if you have concerns or if this happen somewhere "globally".

PTAL:

**Review by Kartikey-star — COMMENTED** [review]
Considering the fact that you are encoding the chart url and index entry values, the backend code does not decode the url encoding. I would suggest that once the https://github.com/openshift/console/pull/11571/ PR goes through we can merge this PR along with decoding done on backend so as to avoid any point of failure.

**Comment by Kartikey-star** [comment]
@debsmita1 as long as the encoding is not changing the charturl and index entry values you proceed with PR. i will be updating the backend in a later PR with decoding values and covering the corner cases with tests .cc @dperaza4dustbit

**Review by jerolimov — CHANGES_REQUESTED** [review]
As discussed in slack, I found two issues:

1. I can select now a helm chart version when installing a helm chart. But I can not select a version anymore when upgrading it.
2. When I add a helm chart and select another version (use Nodejs 0.0.1 instead of 0.0.2) I can not create the helm chart at all.
   Error: `Failed to install helm chart: failed to download ""`
   Looks like the API request doesn't contain `chart_url` in this case

**Comment by jerolimov** [comment]
@Kartikey-star:
> Considering the fact that you are encoding the chart url and index entry values, the backend code does not currently decode the values passed by frontend . I would suggest that once the https://github.com/openshift/console/pull/11571 PR goes through we can merge this PR along with decoding done on backend so as to avoid any point of failure.

> as long as the encoding is not changing the charturl and index entry values you proceed with PR. i will be updating the backend in a later PR with decoding values and covering the corner cases with tests .cc @dperaza4dustbit

Hey @debsmita1, hey @Kartikey-star,

the added encoding only affects the frontend. The code now escapes the URL when the user selects a helm chart from the catalog. This is required because we don't know which characters the URL and build a Link `href` that contains the URL. If we would use a JS API like `URLSearchParams` this wouldn't be required.

It is ***correctly not encoded*** when we pass them as POST call to the API. So there is no change needed in the backend. At least not to decode something. For sure, you still need to consume `indexEntry`.

I have verified this by manually:
* added `?x=1&y=2` in `frontend/packages/helm-plugin/src/catalog/utils/catalog-utils.tsx` to the `chartURL` in `normalizeHelmCharts` 
* check the API call
* and the variables in `pkg/helm/handlers/handlers.go` > `HandleHelmInstall`

File: frontend/packages/helm-plugin/src/components/forms/install-upgrade/HelmInstallUpgradePage.tsx [hunk 5]
```diff
@@ -55,9 +56,10 @@ const HelmInstallUpgradePage: React.FunctionComponent<HelmInstallUpgradePageProp
 
   const namespace = match.params.ns || searchParams.get('preselected-ns');
   const initialChartURL = decodeURIComponent(searchParams.get('chartURL'));
+  const indexEntry = decodeURIComponent(searchParams.get('indexEntry'));
```
Comment (jerolimov) [inline frontend/packages/helm-plugin/src/components/forms/install-upgrade/HelmInstallUpgradePage.tsx:59]: This decodeURIComponent is not needed as URLSearchParams already understand, split and extract the query or search parameters from the URL.

```suggestion
  const indexEntry = searchParams.get('indexEntry');
```

File: frontend/packages/helm-plugin/src/components/forms/install-upgrade/HelmInstallUpgradePage.tsx [hunk 6]
```diff
@@ -55,9 +56,10 @@ const HelmInstallUpgradePage: React.FunctionComponent<HelmInstallUpgradePageProp
...
   const initialReleaseName = match.params.releaseName || '';
-  const helmChartName = searchParams.get('chartName');
-  const helmChartRepoName = searchParams.get('chartRepoName');
+  const helmChartName = decodeURIComponent(searchParams.get('chartName'));
+  const helmChartRepoName = decodeURIComponent(searchParams.get('chartRepoName'));
```
Comment (jerolimov) [inline frontend/packages/helm-plugin/src/components/forms/install-upgrade/HelmInstallUpgradePage.tsx:62]: Same here. `encodeURIComponent` and `decodeURIComponent` is only needed when you build and extract something manually from an URL. When we use URLSearchParams this is not needed.

```suggestion
  const helmChartName = searchParams.get('chartName');
  const helmChartRepoName = searchParams.get('chartRepoName');
```

**Comment by debsmita1** [comment]
> As discussed in slack, I found two issues:
> 
> 1. I can select now a helm chart version when installing a helm chart. But I can not select a version anymore when upgrading it.
> 2. When I add a helm chart and select another version (use Nodejs 0.0.1 instead of 0.0.2) I can not create the helm chart at all.
>    Error: `Failed to install helm chart: failed to download ""`
>    Looks like the API request doesn't contain `chart_url` in this case

I am able to upgrade helm charts without any issues, attached a gif
Uploading Screen Recording 2022-07-26 at 8.26.59 PM.mov…

File: frontend/packages/helm-plugin/src/utils/helm-utils.ts [hunk 7]
```diff
@@ -100,7 +99,28 @@ export const getChartRepositoryTitle = (
   chartRepoName: string,
 ) => {
   const chartRepository = chartRepositories?.find((repo) => repo.metadata.name === chartRepoName);
-  return chartRepository?.spec?.name || toTitleCase(chartRepoName);
+  if (chartRepository) {
+    return chartRepository?.spec?.name;
+  }
```
Comment (jerolimov) [inline frontend/packages/helm-plugin/src/utils/helm-utils.ts:104]: The display name is optional, so this code should fallback to the chartRepoName when it's null, undefined or empty.

```suggestion
  if (chartRepository?.spec?.name) {
    return chartRepository.spec.name;
  }
```

File: frontend/packages/helm-plugin/src/components/forms/install-upgrade/HelmInstallUpgradePage.tsx [hunk 8]
```diff
@@ -74,7 +76,7 @@ const HelmInstallUpgradePage: React.FunctionComponent<HelmInstallUpgradePageProp
   const [initialFormSchema, setInitialFormSchema] = React.useState<JSONSchema7>();
 
   const helmAction: HelmActionType =
-    initialChartURL !== 'null' ? HelmActionType.Install : HelmActionType.Upgrade;
+    initialChartURL !== null ? HelmActionType.Install : HelmActionType.Upgrade;
```
Comment (jerolimov) [inline frontend/packages/helm-plugin/src/components/forms/install-upgrade/HelmInstallUpgradePage.tsx:79]: Can you explain why this is now `null` instead of `"null"`? And can we simplify this?

```suggestion
    !initialChartURL ? HelmActionType.Install : HelmActionType.Upgrade;
```

**Review by jerolimov — CHANGES_REQUESTED** [review]
One small question, and one request to check if the display name is defined or not.

File: frontend/packages/helm-plugin/src/components/forms/install-upgrade/HelmInstallUpgradePage.tsx (hunk 8 above)
Comment (jerolimov) [reply frontend/packages/helm-plugin/src/components/forms/install-upgrade/HelmInstallUpgradePage.tsx]: Thanks for explaining this :+1:

**Review by jerolimov — APPROVED** [review]
Tested this locally: saw the new network API arguments and can now select a helm chart version when installing a helm chart. Upgrade works now also as expected.

/lgtm
/approve
